core/src/main/scala/org/bitcoins/core/script/interpreter/ScriptInterpreter.scala

```diff
@@ -477,18 +545,233 @@ sealed abstract class ScriptInterpreter {
+                    _ == ScriptVerifyDiscourageUpgradableTaprootVersion)
+                if (controlBlock.isTapLeafMask) {
+
+                  //drop the control block & script in the witness
+                  val stackNoControlBlockOrScript = {
+                    if (scriptPubKeyExecutedProgram.getAnnexHashOpt.isDefined) {
+                      //if we have an annex we need to drop
+                      //annex,control block, script
+                      stack.tail.tail.tail
+                    } else {
+                      //else just drop control block, script
+                      stack.tail.tail
+                    }
+                  }
+                  val newProgram = PreExecutionScriptProgram(
+                    txSignatureComponent = taprootTxSigComponent,
+                    stack = stackNoControlBlockOrScript.toList,
+                    script = rebuiltSPK.asm.toList,
+                    originalScript = rebuiltSPK.asm.toList,
+                    altStack = Nil,
+                    flags = taprootTxSigComponent.flags
+                  )
+                  val evaluated = executeProgram(newProgram)
+                  val segwitChecks = postSegWitProgramChecks(evaluated)
+                  Success(segwitChecks)
+                } else if (isDiscouragedTaprootVersion) {
+                  val p = scriptPubKeyExecutedProgram.failExecution(
+                    ScriptErrorDiscourageUpgradableTaprootVersion)
+                  Success(p)
+                } else {
+                  //is this right? I believe to maintain softfork compatibility we
+                  //just succeed?
+                  Success(scriptPubKeyExecutedProgram)
+                }
+              }
+            }
+        }
+      }
+    } else {
+      //if taproot flag not set trivially pass
+      Success(scriptPubKeyExecutedProgram)
+    }
+  }
+
+  private def handleOpSuccess(
+      scriptPubKeyExecutedProgram: ExecutedScriptProgram): Try[
+    ExecutedScriptProgram] = {
+    val discourageOpSuccess =
+      ScriptFlagUtil.discourageOpSuccess(scriptPubKeyExecutedProgram.flags)
+    val p = if (discourageOpSuccess) {
+      scriptPubKeyExecutedProgram.failExecution(ScriptErrorDiscourageOpSuccess)
+    } else {
+      scriptPubKeyExecutedProgram
+    }
+    Success(p)
+  }
+
+  /** Similar to the check schnorr signature method in bitcoin core
+    * @see https://github.com/bitcoin/bitcoin/blob/b71d37da2c8c8d2a9cef020731767a6929db54b4/src/script/interpreter.cpp#L1673
+    */
+  private def checkSchnorrSignature(
+      witness: TaprootKeyPath,
+      pubKey: SchnorrPublicKey,
+      program: ExecutedScriptProgram): ExecutedScriptProgram = {
+    val scriptResult = TransactionSignatureChecker.checkSchnorrSignature(
+      program.txSignatureComponent,
+      pubKey,
+      witness,
+      program.taprootSerializationOptions)
+    scriptResult match {
+      case ScriptOk         => program
```

Comment:
I needed to make this change to get keyspend validation working, otherwise the `ScriptVerifyCleanStack` flag would fail

https://github.com/Christewart/bitcoin-s-core/pull/37/commits/9c1c71128112ca9419873113677683c75df8fc88

Reply:
Done in: https://github.com/bitcoin-s/bitcoin-s/pull/3769/commits/8bca01463761a905a70c8dd486a073af71c2e5f5